src/lib/containers/useImmutableTableQuery.ts

```diff
@@ -0,0 +1,171 @@
+   *  because we could uniformly handle all complex stateful logic in this hook
+   */
+}
+
+export type UseImmutableTableQueryOptions = {
+  /** The initial table query request object */
+  initQueryRequest: QueryBundleRequest
+  /** Whether the URL should update when the query is modified. */
+  shouldDeepLink?: boolean
+  /** Unique index for the component on the page so URL updates do not conflict between table query components */
+  componentIndex?: number
+  /** Callback invoked when the query is modified */
+  onQueryChange?: (newQueryJson: string) => void
+}
+
+/**
+ * Custom hook that maintains and manages the state of a Synapse Table query.
+ * @param options
+ * @returns
+ */
+export default function useImmutableTableQuery(
+  options: UseImmutableTableQueryOptions,
+): ImmutableTableQueryResult {
+  const {
+    initQueryRequest,
+    componentIndex = 0,
+    shouldDeepLink = false,
+    onQueryChange,
+  } = options
+
+  const [lastQueryRequest, setLastQueryRequest] =
+    useState<QueryBundleRequest>(initQueryRequest)
+
+  /**
+   * Inspect the URL on mount to see if we have a particular query request that we must show.
+   */
+  useEffect(() => {
+    const query = DeepLinkingUtils.getQueryRequestFromLink(
+      'QueryWrapper',
+      componentIndex,
+    )
+    if (query) {
+      setLastQueryRequest(query)
+    }
+  }, [componentIndex])
+
+  /**
+   * Pass down a deep clone (so no side effects on the child's part) of the
+   * last query request made
+   *
+   * @returns
+   * @memberof QueryWrapper
+   */
+  const getLastQueryRequest = useCallback(() => {
+    return cloneDeep(lastQueryRequest)
+  }, [lastQueryRequest])
+
+  /**
+   * Pass down a deep clone (so no side affects on the child's part) of the
+   * first query request made
+   *
+   * @returns
+   * @memberof QueryWrapper
+   */
+  function getInitQueryRequest(): QueryBundleRequest {
+    return cloneDeep(initQueryRequest)
+  }
+
+  /**
+   * Execute the given query request, updating all of the data in the QueryContext to match the new query
+   * @param {*} queryRequest Query request as specified by
+   *                         https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/table/Query.html
+   */
+  function setQuery(
+    queryRequest:
+      | QueryBundleRequest
+      | ((lastQueryRequest: QueryBundleRequest) => QueryBundleRequest),
+  ): void {
+    let newQueryRequest: QueryBundleRequest
+    if (typeof queryRequest === 'function') {
+      newQueryRequest = queryRequest(getLastQueryRequest())
+    } else {
+      newQueryRequest = queryRequest
+    }
+
+    // Clone the query request before storing it in state, in case the original is mutated
+    const clonedQueryRequest = cloneDeep(newQueryRequest)
+    setLastQueryRequest(clonedQueryRequest)
+
+    if (clonedQueryRequest.query) {
+      const clonedQueryRequestJson = JSON.stringify(clonedQueryRequest.query)
```

Comment:
I know it was probably doing this before, but it looks like we only need `clonedQueryRequestJson` and `stringifiedQuery` if `shouldDeepLink` is true (so these two lines could be moved into the `if (shouldDeepLink)` block).